packages/compiler-cli/src/ngtsc/metadata/src/util.ts

```diff
@@ -93,7 +96,45 @@ export function extractDirectiveGuards(node: ClassDeclaration, reflector: Reflec
   const coercedInputFields =
       new Set(staticMembers.map(extractCoercedInput)
                   .filter((inputName): inputName is string => inputName !== null));
-  return {hasNgTemplateContextGuard, ngTemplateGuards, coercedInputFields};
+
+  const restrictedInputFields = new Set<string>();
+  const stringLiteralInputFields = new Set<string>();
+  const undeclaredInputFields = new Set<string>();
+
+  for (const fieldName of Object.keys(inputs)) {
+    const field = members.find(member => member.name === fieldName);
+    if (field === undefined || field.node === null) {
+      undeclaredInputFields.add(fieldName);
+    } else if (isRestricted(field.node)) {
+      restrictedInputFields.add(fieldName);
+    }
+    if (field && field.nameNode !== null && ts.isStringLiteral(field.nameNode)) {
+      stringLiteralInputFields.add(fieldName);
+    }
```

Comment:
nit:

```suggestion
    } else {
      if (isRestricted(field.node)) {
        restrictedInputFields.add(fieldName);
      }
      if (ts.isStringLiteral(field.nameNode)) {
        stringLiteralInputFields.add(fieldName);
      }
    }
```

(or maybe even insert a `continue` statement in the first if-block)